src/Synthesiser.cpp

```diff
@@ -289,28 +289,47 @@ void Synthesiser::emitCode(std::ostream& out, const RamStatement& stmt) {
             PRINT_END_COMMENT(out);
         }
 
-        void visitQuery(const RamQuery& insert, std::ostream& out) override {
+        void visitQuery(const RamQuery& query, std::ostream& out) override {
             PRINT_BEGIN_COMMENT(out);
             // enclose operation with a check for an empty relation
             std::set<std::string> inputRelNames;
             std::string projectRelName;
             int projectRelArity = -1;
 
-            if (insert.getCondition() != nullptr) {
-                out << "if(";
-                visit(*insert.getCondition(), out);
-                out << ") {\n";
+            // check whether the outer-most filter operation
+            // can be pushed out of the parallel context
+            const RamOperation* next = &query.getOperation();
+            const RamCondition* cond = nullptr;
+            if (const RamFilter* filter = dynamic_cast<const RamFilter*>(&query.getOperation())) {
+                // cannot pull filter out of parallel loop if it requires a context for
+                // existence checks of tuples
+                cond = &filter->getCondition();
+                bool needContext = false;
+                visitDepthFirst(*cond, [&](const RamExistenceCheck& exists) { needContext = true; });
+                if (!needContext) {
+                    // discharge condition of filter in
+                    // the outer scope
+                    next = &filter->getOperation();
+                    out << "if(";
+                    visit(*cond, out);
+                    out << ") {\n";
+                } else {
+                    // undo filter / contains context operations
+                    cond = nullptr;
+                }
             }
 
-            visitDepthFirst(insert, [&](const RamProject& project) {
+            // get name of projection relation and arity
+            visitDepthFirst(query, [&](const RamProject& project) {
                 projectRelArity = project.getRelation().getArity();
                 projectRelName = project.getRelation().getName();
             });
 
-            visitDepthFirst(insert, [&](const RamRelationSearch& scan) {
+            // TODO: the check for empty relations of a query needs
```

Comment:
If an in-code TODO is needed it should refer to an issue/owner:
`//TODO (#issuenum): ...`
`//TODO (dev username): ...`

Reply:
Done.